File: aws/s3/typescript/src/services/buckets.ts
/**
 * Buckets Service
 *
 * S3 bucket operations: create, delete, head, list, getLocation.
 */

import { S3Config, resolveEndpoint, buildPath } from "../config";
import { S3Error, mapS3ErrorCode } from "../error";
import { AwsSignerV4 } from "../signing";
import { HttpTransport, isSuccess, getHeader, getRequestId } from "../transport";
import {
  CreateBucketRequest,
  DeleteBucketRequest,
  HeadBucketRequest,
} from "../types/requests";
import {
  CreateBucketOutput,
  HeadBucketOutput,
  ListBucketsOutput,
  GetBucketLocationOutput,
} from "../types/responses";
import * as xml from "../xml";

/**
 * Buckets service for S3 bucket operations.
 */
export class BucketsService {
  private config: S3Config;
  private transport: HttpTransport;
  private signer: AwsSignerV4;

  constructor(config: S3Config, transport: HttpTransport, signer: AwsSignerV4) {
    this.config = config;
    this.transport = transport;
    this.signer = signer;
  }

  /**
   * Create a new bucket.
   */
  async create(request: CreateBucketRequest): Promise<CreateBucketOutput> {
    const endpoint = resolveEndpoint(this.config, request.bucket);
    const path = buildPath(this.config, request.bucket);
    const url = new URL(path, endpoint);

    const headers: Record<string, string> = {};

    if (request.acl) {
      headers["x-amz-acl"] = request.acl;
    }
    if (request.grantRead) {
      headers["x-amz-grant-read"] = request.grantRead;
    }
    if (request.grantWrite) {
      headers["x-amz-grant-write"] = request.grantWrite;
    }
    if (request.grantFullControl) {
      headers["x-amz-grant-full-control"] = request.grantFullControl;
    }
    if (request.objectLockEnabled) {
      headers["x-amz-bucket-object-lock-enabled"] = "true";
    }

    let body: Buffer | undefined;

    // Only include location constraint if not us-east-1
    if (request.locationConstraint && request.locationConstraint !== "us-east-1") {
      const xmlBody = xml.buildCreateBucketXml(request.locationConstraint);
      body = Buffer.from(xmlBody);
      headers["content-type"] = "application/xml";
      headers["content-length"] = String(body.length);
    }

    const signed = await this.signer.sign("PUT", url, headers, body);

    const response = await this.transport.send({
      method: "PUT",
      url: signed.url.toString(),
      headers: signed.headers,
      body,
    });

    if (!isSuccess(response)) {
      throw this.parseError(response.body.toString());
    }

    return {
      location: getHeader(response, "location"),
      requestId: getRequestId(response),
    };
  }

  /**
   * Delete a bucket.
   */
  async delete(request: DeleteBucketRequest): Promise<void> {
    const endpoint = resolveEndpoint(this.config, request.bucket);
    const path = buildPath(this.config, request.bucket);
    const url = new URL(path, endpoint);

    const headers: Record<string, string> = {};

    const signed = await this.signer.sign("DELETE", url, headers);

    const response = await this.transport.send({
      method: "DELETE",
      url: signed.url.toString(),
      headers: signed.headers,
    });

    if (!isSuccess(response)) {
      throw this.parseError(response.body.toString());
    }
  }

  /**
   * Check if a bucket exists and is accessible.
   */
  async head(request: HeadBucketRequest): Promise<HeadBucketOutput> {
    const endpoint = resolveEndpoint(this.config, request.bucket);
    const path = buildPath(this.config, request.bucket);
    const url = new URL(path, endpoint);

    const headers: Record<string, string> = {};

    const signed = await this.signer.sign("HEAD", url, headers);

    const response = await this.transport.send({
      method: "HEAD",
      url: signed.url.toString(),
      headers: signed.headers,
    });

    if (!isSuccess(response)) {
      throw this.parseError(response.body.toString());
    }

    return {
      bucketRegion: getHeader(response, "x-amz-bucket-region"),
      accessPointAlias: getHeader(response, "x-amz-access-point-alias") === "true",
      requestId: getRequestId(response),
    };
  }

  /**
   * List all buckets owned by the authenticated sender.
   */
  async list(): Promise<ListBucketsOutput> {
    const endpoint = this.config.endpoint
      ? this.config.endpoint
      : `https://s3.${this.config.region}.amazonaws.com`;
    const url = new URL("/", endpoint);

    const headers: Record<string, string> = {};

    const signed = await this.signer.sign("GET", url, headers);

    const response = await this.transport.send({
      method: "GET",
      url: signed.url.toString(),
      headers: signed.headers,
    });

    if (!isSuccess(response)) {
      throw this.parseError(response.body.toString());
    }

    const result = xml.parseListBuckets(response.body.toString());

    return {
      owner: result.owner,
      buckets: result.buckets,
      requestId: getRequestId(response),
    };
  }

  /**
   * Get bucket location (region).
   */
  async getLocation(bucket: string): Promise<GetBucketLocationOutput> {
    const endpoint = resolveEndpoint(this.config, bucket);
    const path = buildPath(this.config, bucket);
    const url = new URL(path, endpoint);
    url.searchParams.set("location", "");

    const headers: Record<string, string> = {};

    const signed = await this.signer.sign("GET", url, headers);

    const response = await this.transport.send({
      method: "GET",
      url: signed.url.toString(),
      headers: signed.headers,
    });

    if (!isSuccess(response)) {
      throw this.parseError(response.body.toString());
    }

    const location = xml.parseGetBucketLocation(response.body.toString());

    return {
      location: location || "us-east-1", // Empty means us-east-1
      requestId: getRequestId(response),
    };
  }

  private parseError(body: string): S3Error {
    if (!body) {
      return new S3Error("Empty error response", "UnknownError");
    }
    const errorResponse = xml.parseErrorResponse(body);
    return mapS3ErrorCode(errorResponse.code, errorResponse);
  }
}

File: aws/s3/rust/src/services/buckets.rs
//! Buckets service for S3 bucket operations.

use crate::config::S3Config;
use crate::error::{BucketError, S3Error};
use crate::signing::AwsSigner;
use crate::transport::{HttpRequest, HttpTransport};
use crate::types::*;
use crate::xml;
use bytes::Bytes;
use std::collections::HashMap;
use std::sync::Arc;
use url::Url;

/// Service for S3 bucket operations.
pub struct BucketsService {
    config: Arc<S3Config>,
    transport: Arc<dyn HttpTransport>,
    signer: Arc<dyn AwsSigner>,
}

impl BucketsService {
    /// Create a new buckets service.
    pub fn new(
        config: Arc<S3Config>,
        transport: Arc<dyn HttpTransport>,
        signer: Arc<dyn AwsSigner>,
    ) -> Self {
        Self {
            config,
            transport,
            signer,
        }
    }

    /// Create a new bucket.
    pub async fn create(&self, request: CreateBucketRequest) -> Result<CreateBucketOutput, S3Error> {
        let url = self.build_url(Some(&request.bucket), None)?;

        let mut headers = HashMap::new();

        if let Some(acl) = &request.acl {
            headers.insert("x-amz-acl".to_string(), acl.as_str().to_string());
        }

        if let Some(grant_read) = &request.grant_read {
            headers.insert("x-amz-grant-read".to_string(), grant_read.clone());
        }

        if let Some(grant_write) = &request.grant_write {
            headers.insert("x-amz-grant-write".to_string(), grant_write.clone());
        }

        if let Some(grant_full_control) = &request.grant_full_control {
            headers.insert(
                "x-amz-grant-full-control".to_string(),
                grant_full_control.clone(),
            );
        }

        if request.object_lock_enabled == Some(true) {
            headers.insert(
                "x-amz-bucket-object-lock-enabled".to_string(),
                "true".to_string(),
            );
        }

        // Build body with location constraint if not us-east-1
        let region = request
            .location_constraint
            .as_ref()
            .unwrap_or(&self.config.region);

        let body = if region != "us-east-1" {
            let xml = xml::build_create_bucket_xml(region);
            Some(Bytes::from(xml))
        } else {
            None
        };

        if let Some(ref b) = body {
            headers.insert("content-type".to_string(), "application/xml".to_string());
            headers.insert("content-length".to_string(), b.len().to_string());
        }

        let signed = self
            .signer
            .sign("PUT", &url, &headers, body.as_deref().map(|b| b.as_ref()))
            .await?;

        let mut http_request = HttpRequest::new("PUT", signed.url.as_str())
            .with_headers(signed.headers);

        if let Some(b) = body {
            http_request = http_request.with_body(b);
        }

        let response = self.transport.send(http_request).await?;

        if !response.is_success() {
            return Err(self.parse_error(&response.body, &request.bucket).await);
        }

        Ok(CreateBucketOutput {
            location: response.get_header("location").map(String::from),
            request_id: response.request_id().map(String::from),
        })
    }

    /// Delete a bucket.
    pub async fn delete(&self, request: DeleteBucketRequest) -> Result<(), S3Error> {
        let url = self.build_url(Some(&request.bucket), None)?;
        let headers = HashMap::new();

        let signed = self.signer.sign("DELETE", &url, &headers, None).await?;

        let http_request = HttpRequest::new("DELETE", signed.url.as_str())
            .with_headers(signed.headers);

        let response = self.transport.send(http_request).await?;

        if !response.is_success() {
            return Err(self.parse_error(&response.body, &request.bucket).await);
        }

        Ok(())
    }

    /// Check if a bucket exists (HEAD bucket).
    pub async fn head(&self, request: HeadBucketRequest) -> Result<HeadBucketOutput, S3Error> {
        let url = self.build_url(Some(&request.bucket), None)?;
        let headers = HashMap::new();

        let signed = self.signer.sign("HEAD", &url, &headers, None).await?;

        let http_request = HttpRequest::new("HEAD", signed.url.as_str())
            .with_headers(signed.headers);

        let response = self.transport.send(http_request).await?;

        if response.status == 404 {
            return Err(S3Error::Bucket(BucketError::NotFound {
                bucket: request.bucket.clone(),
                request_id: response.request_id().map(String::from),
            }));
        }

        if !response.is_success() {
            return Err(self.parse_error(&response.body, &request.bucket).await);
        }

        Ok(HeadBucketOutput {
            bucket_region: response.get_header("x-amz-bucket-region").map(String::from),
            access_point_alias: response
                .get_header("x-amz-access-point-alias")
                .map(|v| v == "true"),
            request_id: response.request_id().map(String::from),
        })
    }

    /// List all buckets.
    pub async fn list(&self) -> Result<ListBucketsOutput, S3Error> {
        let url = self.build_url(None, None)?;
        let headers = HashMap::new();

        let signed = self.signer.sign("GET", &url, &headers, None).await?;

        let http_request = HttpRequest::new("GET", signed.url.as_str())
            .with_headers(signed.headers);

        let response = self.transport.send(http_request).await?;

        if !response.is_success() {
            return Err(self.parse_error(&response.body, "").await);
        }

        let body_str = String::from_utf8_lossy(&response.body);
        let mut output = xml::parse_list_buckets(&body_str)?;
        output.request_id = response.request_id().map(String::from);

        Ok(output)
    }

    /// Get the location (region) of a bucket.
    pub async fn get_location(&self, bucket: &str) -> Result<String, S3Error> {
        let url = self.build_url(Some(bucket), Some("location"))?;
        let headers = HashMap::new();

        let signed = self.signer.sign("GET", &url, &headers, None).await?;

        let http_request = HttpRequest::new("GET", signed.url.as_str())
            .with_headers(signed.headers);

        let response = self.transport.send(http_request).await?;

        if !response.is_success() {
            return Err(self.parse_error(&response.body, bucket).await);
        }

        // Parse location from response
        let body_str = String::from_utf8_lossy(&response.body);

        // The response is XML like: <LocationConstraint>region</LocationConstraint>
        // If empty, the bucket is in us-east-1
        if body_str.contains("<LocationConstraint/>") || body_str.contains("<LocationConstraint></LocationConstraint>") {
            return Ok("us-east-1".to_string());
        }

        // Extract region from XML
        if let Some(start) = body_str.find("<LocationConstraint>") {
            let start = start + 20; // Length of "<LocationConstraint>"
            if let Some(end) = body_str[start..].find("</LocationConstraint>") {
                return Ok(body_str[start..start + end].to_string());
            }
        }

        // Default to us-east-1 if we can't parse
        Ok("us-east-1".to_string())
    }

    /// Check if a bucket exists.
    pub async fn exists(&self, bucket: &str) -> Result<bool, S3Error> {
        match self.head(HeadBucketRequest::new(bucket)).await {
            Ok(_) => Ok(true),
            Err(S3Error::Bucket(BucketError::NotFound { .. })) => Ok(false),
            Err(e) => Err(e),
        }
    }

    fn build_url(&self, bucket: Option<&str>, query: Option<&str>) -> Result<Url, S3Error> {
        let endpoint = self.config.resolve_endpoint(bucket);

        let url_str = if let Some(bucket) = bucket {
            let path = self.config.build_path(bucket, None);
            if let Some(q) = query {
                format!("{}{}?{}", endpoint.as_str().trim_end_matches('/'), path, q)
            } else {
                format!("{}{}", endpoint.as_str().trim_end_matches('/'), path)
            }
        } else {
            endpoint.to_string()
        };

        Url::parse(&url_str).map_err(|e| {
            S3Error::Request(crate::error::RequestError::Validation {
                message: format!("Invalid URL: {}", e),
            })
        })
    }

    async fn parse_error(&self, body: &Bytes, bucket: &str) -> S3Error {
        if body.is_empty() {
            return S3Error::Response(crate::error::ResponseError::InvalidResponse {
                message: "Empty error response".to_string(),
            });
        }

        let body_str = String::from_utf8_lossy(body);
        match xml::parse_error_response(&body_str) {
            Ok(mut error_response) => {
                if error_response.bucket.is_none() && !bucket.is_empty() {
                    error_response.bucket = Some(bucket.to_string());
                }
                crate::error::map_s3_error_code(&error_response.code, Some(error_response))
            }
            Err(_) => S3Error::Response(crate::error::ResponseError::InvalidResponse {
                message: format!(
                    "Failed to parse error response: {}",
                    body_str.chars().take(100).collect::<String>()
                ),
            }),
        }
    }
}

impl std::fmt::Debug for BucketsService {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("BucketsService")
            .field("config", &self.config)
            .finish_non_exhaustive()
    }
}
